Language docs (#314)
powdr-org/powdr: LANGUAGE.md

# Powdr Assembly (Powdr-Asm)

Powdr-Asm is used to define instruction sets and write programs that depend on
control flow, with the aim of generating ZKPs about them.
Powdr-Asm follows the execution trace table model shared by many proof systems.
In fact, Powdr-Asm extends Polygon Hermez' [PIL and
zkASM](https://zkevm.polygon.technology/docs/category/polynomial-identity-language).
In such model, the execution is represented by a two-dimensional fixed-size
table, where each column represents a piece of state (for example registers or
memory) and each row represents a step in time for a given execution.

> The program is represented by a transition system that specifies how the state
> evolves in time in the form of constraints.

Reviewer: ```suggestion
Constraints specify how the state evolves from any step to the next.
```

As an example, let's apply that model conceptually to a program that computes
the factorial of a number. Our state consists of two registers:

- CNT - a counter
- ACC - the answer accumulator

> The first row will represent the initial state of our program, that is, `PC =

Reviewer: ```suggestion
The first row represents the initial state of our program, that is, `PC =
```

0`, `CNT = <input>`, `ACC = 1`.
From then on, we compute the subsequent rows with the following pseudocode:

> 1. START:

Reviewer 2: I think numbering the program lines confuses because people could think that the program lines correspond to the table rows.

2. Is CNT = 0?
3. If `yes` go to `END`
4. ACC := ACC * CNT
5. CNT := CNT - 1
6. Go to START
7. END:
8. ACC equals `factorial(input)`

We can compute the following execution trace for the specific input 5:

| CNT | ACC |
|-----|-----|
| 5   | 1   |
| 4   | 5   |
| 3   | 20  |
| 2   | 60  |
| 1   | 120 |
| 0   | **120** |

Note that the size of the table determines how long the trace can be.

## ISA

There are three constructs that can be used to define an instruction set:

- Registers
- Macros
- Instructions

### Registers

Registers are basic state components that can store a single numbers element
each. There are three types of registers:

- Program Counter (PC). Used for control flow.

> - Assignment registers. Used by parametric instructions.

Reviewer 2: ```suggestion
- Assignment registers. Used by parametric instructions and for assignments.
```

- General registers. Used by program logic.

The registers above can be declared in the following ways:

```asm
reg pc[@pc];
reg X[<=];
reg A;
```

In the snippet above, the `@pc` annotation means that `pc` is implicitly
incremented by one for instructions that do not use it.  The `<=` annotation

> means `X` is an assignment register and can be used by parametric instructions.

Reviewer 2: ```suggestion
means `X` is an assignment register and can be used by parametric instructions and for assignments.
```

> Finally, `A` is a general purpose register and can be used by program logic.

Author: Agree, that's also how I modeled it in the ZoKrates Nova concept

### Instructions

> Instructions are predicates that define constraints that must hold when the

Reviewer 2: ```suggestion
Instructions are predicates that define constraints that are enforced only when the
```

instruction is invoked. The constraints may refer to registers on the execution

> trace line that it is invoked and/or the next line.

Reviewer 2: but the next line can only be referenced in equations of the form `x' = ...`, where the RHS is `'`-free. Maybe we should add that just as an extension after explaining the general mechanism?

Author: That's why I wanted to be vague here. I think after Powdr too we'll need a more structured version of this, that's why I'm not sure about this PR in general.

```asm
instr ADD X, Y -> Z { X + Y = Z }
```

The above instruction can be read as "If `Z` is the same as `X + Y`, then `ADD
X Y` returns `Z`".
As an analogy, this is similar to Prolog predicates.

> 

Reviewer 2: If an instruction has formal parameters, these must either be literals or assignment registers.

### Macros

Macros can be used to build expressions which are inlined at the call site.

```asm
macro if_then_else(condition, true_value, false_value) {
    condition * true_value + (1 - condition) * false_value
};
```

The macro above can be used to simplify if-then-else arithmetic expressions.

### Factorial, again

We now have enough constructs to create a simple ISA and implement our
Factorial program.

> ```asm

Author: let's leave it like this for now and adjust after powdr2 is merged

pil {
    col witness XInv;
    col witness XIsZero;
    XIsZero * (1 - XIsZero) = 0;
    XIsZero = 1 - X * XInv;
    XIsZero * X = 0;
}

macro if_then_else(condition, true_value, false_value) {
    condition * true_value + (1 - condition) * false_value
};

> macro jump_to(target) { pc' = target; };

Reviewer 2: Now the big question: Do we require `pc` to be a parameter?

Reviewer: Yes please

Author: I'll leave it like this in this file for now

macro jump_to_if(condition, target) {
    jump_to(if_then_else(condition, target, pc + 1));
};

reg pc[@pc];
reg X[<=];
reg CNT;
reg ACC;

instr jmpz X, l: label { jump_to_if(XIsZero, l) }
instr jmp l: label { jump_to(l) }
instr dec_CNT { CNT' = CNT - 1 }
instr assert_zero X { XIsZero = 1 }

CNT <=X= ${ ("input", 0) };

start::
 jmpz CNT, end;
 ACC <=X= ACC + ${ ("input", CNT) };
 dec_CNT;
 jmp start;

end::
```

> The beginning of the snippet above is necessary to compute whether `X` is 0 at

Reviewer 2: ```suggestion
The beginning of the snippet above is necessary gadget that computes whether `X` is 0 at
```
